### PR TITLE
Reduce the initial JS size

### PR DESCRIPTION
**Pull request by jonkafton**
### What are the relevant tickets?

Closes https://github.com/mitodl/hq/issues/6224

### Description (What does it do?)

Reduces the size of bundled JavaScript chunks needed to hydrate and bring interactivity to the page. This does not improve the initial render time (the server rendered pages are displayed and we have the DOMContentLoaded event before the scripts have downloaded), but aims to reduce the payload size of individual JS chunks by enabling the bundler to distribute code across chunks by tree shaking and thereby allowing for only the code in use on a page to be loaded.

- Tree shakes ol-components (and therefore MUI)
 
- Lazy loads the Resource Drawer

- Tree shakes Lodash

- Also removes ol-ckeditor and ol-widgets as dead code. These were not making their way to the client bundle, though needed some work to be compatible with the removal of the ol-components index file - arguably not worth doing unless we reinstate them.

### How can this be tested?

Run `ANALYZE=true yarn build` in the main workspace.

Next.js outputs some bundle state by route. We're interested in the "First Load JS" size for each route. I've merged the before and after below:

```
Route (app)                              Size     First Load JS
┌ ƒ /                                    10.3 kB         607 kB
┌ ƒ /                                    16.2 kB         555 kB

├ ○ /_not-found                          304 B           168 kB
├ ○ /_not-found                          307 B           168 kB

├ ○ /about                               3.88 kB         475 kB
├ ○ /about                               6.16 kB         244 kB

├ ƒ /c/[channelType]/[name]              8.33 kB         623 kB
├ ƒ /c/[channelType]/[name]              17.2 kB         609 kB

├ ○ /cart                                2.02 kB         522 kB
├ ○ /cart                                4.37 kB         286 kB

├ ○ /dashboard                           336 B           580 kB
├ ○ /dashboard                           396 B           583 kB

├ ƒ /dashboard/[tab]                     335 B           580 kB
├ ƒ /dashboard/[tab]                     397 B           583 kB

├ ƒ /dashboard/[tab]/[id]                336 B           580 kB
├ ƒ /dashboard/[tab]/[id]                396 B           583 kB

├ ○ /departments                         4.4 kB          492 kB
├ ○ /departments                         9.68 kB         369 kB

├ ○ /learningpaths                       3.72 kB         520 kB
├ ○ /learningpaths                       3.05 kB         458 kB

├ ƒ /learningpaths/[id]                  2.77 kB         575 kB
├ ƒ /learningpaths/[id]                  4.07 kB         536 kB

├ ○ /onboarding                          3.65 kB         517 kB
├ ○ /onboarding                          16 kB           359 kB

├ ○ /privacy                             5.24 kB         477 kB
├ ○ /privacy                             7.54 kB         240 kB

├ ƒ /program_letter/[id]/view            2.19 kB         485 kB
├ ƒ /program_letter/[id]/view            2.37 kB         218 kB

├ ƒ /search                              1.67 kB         617 kB
├ ƒ /search                              1.78 kB         577 kB

├ ○ /terms                               4.21 kB         476 kB
├ ○ /terms                               6.52 kB         239 kB

├ ○ /topics                              4.69 kB         493 kB
├ ○ /topics                              7.86 kB         371 kB

└ ○ /units                               4.46 kB         492 kB
└ ○ /units                               14.1 kB         359 kB
```

The `@next/bundle-analyzer` will open 3 html reports. We're interested in the third, client.html. I suggest selecting "Gzipped" in the left menu so we're consistent on numbers; as everything is served compressed.

The before shows that MUI is the single largest library offender. We're not loading all of MUI, as that itself is tree shaken as we import it into ol-components, however ol-components were loaded through a single entrypoint, forcing it to be loaded in a single chunk:

image

<br />

After tree shaking ol-components so that MUI can be distributed across chunks:

image

Lodash also stood out - it's small utility code barely in use. Tree shaking that eliminated the majority of its footprint.

It's interesting to filter on chunks for an individual route and to search for specific modules. Filtering on app/page (the homepage) for "MUI", we can see how this has been distributed across chunks in the before and after:

image

vs.

image

### Additional Context

The initial plan was to focus on lazy loading portions of our code that are not needed until the user asks for them, the dialog modals in particular. I've done that for the Resource Drawer, though that produced a grand cost saving of 607kB -> 606 kB. It turns out that library code not tree shaken is by far the biggest offender. In the absence of any utility code, lazy loading the Resource Drawer looks like: 

```
const LearningResourceDrawer = dynamic(
  () =>
    import("@/page-components/LearningResourceDrawer/LearningResourceDrawer"),
)
```

The `next/dynamic` call only supports React components, so we cannot use it for plain functions for example for any named export hooks imported alongside components. It does not seem to be able to infer the types, so we'd need to rework the code to import types. In turns out that relative to the effort and added code complexity, the benefits of lazy loading bits of our own code our very small, whereas by tree shaking our library code (MUI in particular) so the bundler isn't forced to insert it into a single chunk, the Next.js bundler seems to do a good job of distributing code into chunks that are not needed for initial interactivity. 

The bulk of this PR is to remove the ol-components index and import components directly so that it (and therefore MUI) is tree shaken, with the added incentive that we are migrating to the Smoot Design library and are no longer re-exporting MUI components there, so would need to update the Learn code in this direction. We have the **action**  for the Smoot Design library to ensure that is is tree shakable, though should be able to achieve by marking as [side-effect-free](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) so that we can still import in the the form `import { Button } from "@mitodl/smoot-design"`.

Another goal aside from simply reducing the JS payload is to improve the Total Blocking Time performance metric by eliminating code that may block the execution thread. It may be that there are gains to be had there by lazy loading more of our app code, though I haven't seen much evidence of that while inspecting the dev tools profiler. We will need to investigate that further.

Other libraries for lazy loading:

- Posthog is now the largest single chunk on several pages. It doesn't offer support for lazy loading, but it would be on the list to target if we want to optimize further. I spent some time looking at this, but needed some rework (async imports inside sync FCs, `usePostHog()` is a hook) and was getting bloated.

- Moment is another library loaded in its entirety. We can't tree shake that as dates need to be passed to its default export.

### CODE REVIEW

File: frontends/main/src/types/typography.d.ts
Comment (jonkafton) [inline frontends/main/src/types/typography.d.ts:1]: I didn't have luck getting this to be available with the TS config. Copied over for now, though they shouldn't be here.

**Comment by ChristopherChudzicki** [comment]
> ... however ol-components were loaded through a single entrypoint, forcing it to be loaded in a single chunk:

This should not prevent tree-shaking. For example
- https://mui.com/material-ui/guides/minimizing-bundle-size/#when-and-how-to-use-tree-shaking
- https://stackoverflow.com/a/64569600

However, I did come across this [optimizePackageImports](https://nextjs.org/docs/pages/api-reference/next-config-js/optimizePackageImports) the other day, which leads me to believe NextJS isn't doing a great job handling this.

**NB:** The reason we import from individual mui packages is not tree shaking, but the fact that "barrel" files can cause CJS-based things to be very slow. (e.g., https://github.com/jestjs/jest/issues/11234, https://mui.com/material-ui/guides/minimizing-bundle-size/#development-environment)

**Comment by jonkafton** [comment]
> Edit: Actually, I don't think we need optimizePackageImports... I think adding "sideEffects": false to ol-components package.json should help.

Yes! Adding `"sideEffects": false` to ol-components, without the [commit](https://github.com/mitodl/mit-learn/pull/1901/commits/5d29ed4ca50baa636743c1d842a898a5a7f2abee) to remove the ol-components index and use path imports tree-shakes MUI and actually gave better results:

```
Route (app)                              Size     First Load JS
┌ ƒ /                                    17.1 kB         532 kB
├ ○ /_not-found                          307 B           168 kB
├ ○ /about                               3.54 kB         230 kB
├ ƒ /c/[channelType]/[name]              10.5 kB         580 kB
├ ○ /cart                                1.68 kB         272 kB
├ ○ /dashboard                           398 B           558 kB
├ ƒ /dashboard/[tab]                     397 B           558 kB
├ ƒ /dashboard/[tab]/[id]                395 B           558 kB
├ ○ /departments                         10.7 kB         353 kB
├ ○ /learningpaths                       6.46 kB         440 kB
├ ƒ /learningpaths/[id]                  4.95 kB         516 kB
├ ○ /onboarding                          16.1 kB         344 kB
├ ○ /privacy                             4.88 kB         226 kB
├ ƒ /program_letter/[id]/view            2.37 kB         217 kB
├ ƒ /search                              5.58 kB         552 kB
├ ○ /terms                               3.85 kB         225 kB
├ ○ /topics                              7.54 kB         355 kB
└ ○ /units                               10.4 kB         345 kB
```

As @ChristopherChudzicki  suggested, adding `"sideEffects": false` to ol-utilities improves further:
```
Route (app)                              Size     First Load JS
┌ ƒ /                                    17.2 kB         491 kB
├ ○ /_not-found                          307 B           168 kB
├ ○ /about                               3.54 kB         230 kB
├ ƒ /c/[channelType]/[name]              52 kB           580 kB
├ ○ /cart                                1.68 kB         272 kB
├ ○ /dashboard                           393 B           517 kB
├ ƒ /dashboard/[tab]                     394 B           517 kB
├ ƒ /dashboard/[tab]/[id]                393 B           517 kB
├ ○ /departments                         9.52 kB         309 kB
├ ○ /learningpaths                       6.45 kB         399 kB
├ ƒ /learningpaths/[id]                  5.01 kB         475 kB
├ ○ /onboarding                          16.1 kB         343 kB
├ ○ /privacy                             4.88 kB         226 kB
├ ƒ /program_letter/[id]/view            2.37 kB         217 kB
├ ƒ /search                              5.57 kB         511 kB
├ ○ /terms                               3.85 kB         225 kB
├ ○ /topics                              5.42 kB         289 kB
└ ○ /units                               10.7 kB         279 kB
```

With the removal of ol-components barrel file and `"sideEffects": false`, the results are... worse! :exploding_head:

```
Route (app)                              Size     First Load JS
┌ ƒ /                                    18.7 kB         517 kB
├ ○ /_not-found                          307 B           168 kB
├ ○ /about                               5.15 kB         245 kB
├ ƒ /c/[channelType]/[name]              52 kB           609 kB
├ ○ /cart                                3.72 kB         287 kB
├ ○ /dashboard                           395 B           547 kB
├ ƒ /dashboard/[tab]                     396 B           547 kB
├ ƒ /dashboard/[tab]/[id]                394 B           547 kB
├ ○ /departments                         11.8 kB         326 kB
├ ○ /learningpaths                       3.04 kB         416 kB
├ ƒ /learningpaths/[id]                  4.15 kB         494 kB
├ ○ /onboarding                          16.1 kB         363 kB
├ ○ /privacy                             6.52 kB         241 kB
├ ƒ /program_letter/[id]/view            2.3 kB          218 kB
├ ƒ /search                              5.55 kB         535 kB
├ ○ /terms                               5.48 kB         240 kB
├ ○ /topics                              6.9 kB          303 kB
└ ○ /units                               11.5 kB         294 kB
```

I have a small preference for path + default imports for component libraries, ie. `import Component from "library/Component"` vs. `import { Component } from "library"` - it doesn't depend on configuration for the bundler and prevents naming collisions (however unavoidable).

Given these results though, the change to remove ol-components/src/index.ts was unnecessary. I've switched the branch for this PR to `jk/6224-reduce-initial-js-2`, though we have the original branch at `jk/6224-reduce-initial-js` if we decide that path imports are an improvement stylistically, or if we see performance improvement for the tests or elsewhere. The tests btw timed at 29.673s with pathed imports and 30.904s with the barrel file, so no apparent difference locally.

File: frontends/package.json [hunk 1]
```diff
@@ -4,13 +4,11 @@
   "workspaces": [
     "api",
     "main",
-    "ol-ckeditor",
```
Comment (ChristopherChudzicki) [inline frontends/package.json:7]: OK, if we need it, git has it :)

**Review by ChristopherChudzicki — APPROVED** [review]
@jonkafton Happy to approve this as the "First Load JS" decrease does seem very good.

That said, we do seem to be downloading more than just the first load js:

Screenshot 2024-12-17 at 12 52 21 PM

whereas first load js is 496kb.

My suspicious is that:
- first laod JS = script tags included in HTML do (excluding `async` script tags)
- the initial js triggers some extra loads that might not be needed immediately

**Pull request merged**